File: sdks/ts/memory/src/ontology/validation.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Ontology type validation functions. Validates node types, edge types,
 * business categories, and type definitions against the naming conventions.
 *
 * Port of packages/intelligence-database validation logic.
 */

import type { OntologyTypeDefinition, TypeStatus } from './types.js'
import { BUILT_IN_EDGE_TYPES, BUILT_IN_NODE_TYPES, BUSINESS_CATEGORIES, NODE_TYPE_PREFIXES } from './types.js'

const SNAKE_CASE_NAME_RE = /^[a-z][a-z0-9]*(_[a-z0-9]+)*$/

const builtInNodeTypeSet: ReadonlySet<string> = new Set(BUILT_IN_NODE_TYPES)
const builtInEdgeTypeSet: ReadonlySet<string> = new Set(BUILT_IN_EDGE_TYPES)
const businessCategorySet: ReadonlySet<string> = new Set(BUSINESS_CATEGORIES)

const VALID_STATUSES: ReadonlySet<TypeStatus> = new Set(['active', 'proposed', 'deprecated'])

/**
 * Reports whether value exactly matches a built-in node type.
 */
export function isBuiltInNodeType(value: string): boolean {
  return builtInNodeTypeSet.has(value)
}

/**
 * Reports whether value exactly matches a built-in edge type.
 */
export function isBuiltInEdgeType(value: string): boolean {
  return builtInEdgeTypeSet.has(value)
}

/**
 * Reports whether value exactly matches a built-in business category.
 */
export function isBuiltInBusinessCategory(value: string): boolean {
  return businessCategorySet.has(value)
}

/**
 * Reports whether value is a valid node type identifier.
 * A value is valid if it exactly matches a built-in node type, or if it
 * starts with a valid prefix and has a valid snake_case name after the dot.
 */
export function isValidNodeType(value: string): boolean {
  if (builtInNodeTypeSet.has(value)) {
    return true
  }
  for (const prefix of NODE_TYPE_PREFIXES) {
    if (value.startsWith(prefix)) {
      const name = value.slice(prefix.length)
      if (name.length === 0) {
        return false
      }
      return SNAKE_CASE_NAME_RE.test(name)
    }
  }
  return false
}

/**
 * Reports whether value is a valid edge type identifier.
 * A value is valid if it exactly matches a built-in edge type, or if it
 * matches the snake_case pattern.
 */
export function isValidEdgeType(value: string): boolean {
  if (builtInEdgeTypeSet.has(value)) {
    return true
  }
  return SNAKE_CASE_NAME_RE.test(value)
}

/**
 * Reports whether value is a valid business category identifier.
 */
export function isValidBusinessCategory(value: string): boolean {
  return SNAKE_CASE_NAME_RE.test(value)
}

/**
 * Returns the matching prefix for a node type, or undefined if none matches.
 */
export function hasPrefix(nodeType: string): string | undefined {
  for (const prefix of NODE_TYPE_PREFIXES) {
    if (nodeType.startsWith(prefix)) {
      return prefix
    }
  }
  return undefined
}

/**
 * Returns an error message if value is not a valid node type, or undefined if valid.
 */
export function validateNodeType(value: string): string | undefined {
  if (isValidNodeType(value)) {
    return undefined
  }
  return `ontology: invalid node type "${value}": must start with a valid prefix (entity., rule., exception., decision., process.) followed by a snake_case name`
}

/**
 * Returns an error message if value is not a valid edge type, or undefined if valid.
 */
export function validateEdgeType(value: string): string | undefined {
  if (isValidEdgeType(value)) {
    return undefined
  }
  return `ontology: invalid edge type "${value}": must be lowercase snake_case starting with a letter`
}

/**
 * Returns an error message if value is not a valid business category, or undefined if valid.
 */
export function validateBusinessCategory(value: string): string | undefined {
  if (isValidBusinessCategory(value)) {
    return undefined
  }
  return `ontology: invalid business category "${value}": must be lowercase snake_case starting with a letter`
}

/**
 * Validates a type definition has all required fields and valid values.
 * Returns an error message or undefined if valid.
 */
export function validateTypeDefinition(def: OntologyTypeDefinition): string | undefined {
  if (!def.type) {
    return 'ontology: type definition has empty type field'
  }
  if (!def.label) {
    return `ontology: type definition "${def.type}" has empty label`
  }
  if (!def.description) {
    return `ontology: type definition "${def.type}" has empty description`
  }
  if (!def.createdAt) {
    return `ontology: type definition "${def.type}" has empty createdAt`
  }
  if (!VALID_STATUSES.has(def.status)) {
    return `ontology: type definition "${def.type}" has invalid status "${def.status}": must be active, proposed, or deprecated`
  }
  return undefined
}

File: go/ontology/validation.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import (
	"fmt"
	"regexp"
	"strings"
)

// snakeCaseNameRe matches a valid snake_case name: starts with lowercase letter,
// followed by lowercase alphanumeric segments separated by underscores.
var snakeCaseNameRe = regexp.MustCompile(`^[a-z][a-z0-9]*(_[a-z0-9]+)*$`)

// validStatuses is the set of acceptable TypeStatus values for type definitions.
var validStatuses = map[TypeStatus]struct{}{
	TypeStatusActive:     {},
	TypeStatusProposed:   {},
	TypeStatusDeprecated: {},
}

// IsValidNodeType reports whether value is a valid node type identifier.
// A value is valid if it exactly matches a built-in node type, or if it
// starts with a valid prefix and has a valid snake_case name after the dot.
func IsValidNodeType(value string) bool {
	if IsBuiltInNodeType(value) {
		return true
	}
	prefixMu.RLock()
	prefixes := nodeTypePrefixes
	prefixMu.RUnlock()
	for _, prefix := range prefixes {
		if strings.HasPrefix(value, prefix) {
			name := value[len(prefix):]
			if len(name) == 0 {
				return false
			}
			return snakeCaseNameRe.MatchString(name)
		}
	}
	return false
}

// IsValidEdgeType reports whether value is a valid edge type identifier.
// A value is valid if it exactly matches a built-in edge type, or if it
// matches the snake_case pattern ^[a-z][a-z0-9]*(_[a-z0-9]+)*$.
func IsValidEdgeType(value string) bool {
	if IsBuiltInEdgeType(value) {
		return true
	}
	return snakeCaseNameRe.MatchString(value)
}

// IsValidBusinessCategory reports whether value is a valid business category.
// A value is valid if it matches the snake_case pattern.
func IsValidBusinessCategory(value string) bool {
	return snakeCaseNameRe.MatchString(value)
}

// HasPrefix reports whether the given node type starts with one of the
// known prefixes and returns the prefix. Returns empty string if no prefix matches.
func HasPrefix(nodeType string) string {
	prefixMu.RLock()
	prefixes := nodeTypePrefixes
	prefixMu.RUnlock()
	for _, prefix := range prefixes {
		if strings.HasPrefix(nodeType, prefix) {
			return prefix
		}
	}
	return ""
}

// ValidateNodeType returns an error if value is not a valid node type.
func ValidateNodeType(value string) error {
	if IsValidNodeType(value) {
		return nil
	}
	return fmt.Errorf("ontology: invalid node type %q: must start with a valid prefix (entity., rule., exception., decision., process.) followed by a snake_case name", value)
}

// ValidateEdgeType returns an error if value is not a valid edge type.
func ValidateEdgeType(value string) error {
	if IsValidEdgeType(value) {
		return nil
	}
	return fmt.Errorf("ontology: invalid edge type %q: must be lowercase snake_case starting with a letter", value)
}

// ValidateBusinessCategory returns an error if value is not a valid business category.
func ValidateBusinessCategory(value string) error {
	if IsValidBusinessCategory(value) {
		return nil
	}
	return fmt.Errorf("ontology: invalid business category %q: must be lowercase snake_case starting with a letter", value)
}

// ValidateTypeDefinition checks that a TypeDefinition has all required fields
// and valid values.
func ValidateTypeDefinition(def TypeDefinition) error {
	if def.Type == "" {
		return fmt.Errorf("ontology: type definition has empty type field")
	}
	if def.Label == "" {
		return fmt.Errorf("ontology: type definition %q has empty label", def.Type)
	}
	if def.Description == "" {
		return fmt.Errorf("ontology: type definition %q has empty description", def.Type)
	}
	if def.CreatedAt == "" {
		return fmt.Errorf("ontology: type definition %q has empty createdAt", def.Type)
	}
	if _, ok := validStatuses[def.Status]; !ok {
		return fmt.Errorf("ontology: type definition %q has invalid status %q: must be active, proposed, or deprecated", def.Type, def.Status)
	}
	return nil
}
